### PR TITLE
[ValueTracking] Add support for `trunc nuw/nsw` in isKnowNonZero

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[ValueTracking] Add tests for isKnowNonZero of `trunc nuw/nsw`; NFC**
- **[ValueTracking] Add support for `trunc nuw/nsw` in isKnowNonZero**

With `nsw`/`nuw`, the `trunc` is non-zero if its operand is non-zero.

Proofs: https://alive2.llvm.org/ce/z/iujmk6

### CODE REVIEW

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -3785,7 +3792,8 @@ static unsigned ComputeNumSignBitsImpl(const Value *V,
       // truncation, then we can make use of that. Otherwise we don't know
       // anything.
       Tmp = ComputeNumSignBits(U->getOperand(0), Depth + 1, Q);
-      unsigned OperandTyBits = U->getOperand(0)->getType()->getScalarSizeInBits();
+      unsigned OperandTyBits =
+          U->getOperand(0)->getType()->getScalarSizeInBits();
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:3796]: Unrelated formatting change?

**Review by nikic — APPROVED** [review]
LGTM

**Review by dtcxzyw — CHANGES_REQUESTED** [review]
This patch causes opt to crash :(
I will provide a reproducer later.

**Comment by dtcxzyw** [comment]
```
; opt -passes=licm test.ll -S
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

define void @_ZNK6icu_755Norms7reorderERNS_13UnicodeStringERNS_23BuilderReorderingBufferE(ptr %buffer) {
entry:
  br label %while.cond

while.cond:                                       ; preds = %_ZN6icu_7523BuilderReorderingBuffer6appendEih.exit, %entry
  %0 = load i32, ptr %buffer, align 4
  br i1 false, label %land.rhs.i, label %_ZN6icu_7523BuilderReorderingBuffer6appendEih.exit

land.rhs.i:                                       ; preds = %while.cond
  %1 = trunc nsw i64 0 to i32
  br label %_ZN6icu_7523BuilderReorderingBuffer6appendEih.exit

_ZN6icu_7523BuilderReorderingBuffer6appendEih.exit: ; preds = %land.rhs.i, %while.cond
  %i.0.in.lcssa.i = phi i32 [ %1, %land.rhs.i ], [ %0, %while.cond ]
  %idxprom38.i = sext i32 %i.0.in.lcssa.i to i64
  %arrayidx39.i = getelementptr [31 x i32], ptr %buffer, i64 0, i64 %idxprom38.i
  store i32 0, ptr %arrayidx39.i, align 4
  br label %while.cond
}
```
```
opt: /home/dtcxzyw/WorkSpace/Projects/compilers/llvm-project/llvm/lib/Analysis/ValueTracking.cpp:3024: bool isKnownNonZero(const llvm::Value*, const llvm::APInt&, const llvm::SimplifyQuery&, unsigned int): Assertion `Depth <= MaxAnalysisRecursionDepth && "Limit Search Depth"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: bin/opt -passes=licm reduced.ll
 #0 0x00007f3fd9417a60 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMSupport.so.19.0git+0x217a60)
 #1 0x00007f3fd9414a6f llvm::sys::RunSignalHandlers() (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMSupport.so.19.0git+0x214a6f)
 #2 0x00007f3fd9414bc5 SignalHandler(int) Signals.cpp:0:0
 #3 0x00007f3fd8e42520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007f3fd8e969fc __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x00007f3fd8e969fc __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x00007f3fd8e969fc pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x00007f3fd8e42476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x00007f3fd8e287f3 abort ./stdlib/abort.c:81:7
 #9 0x00007f3fd8e2871b _nl_load_domain ./intl/loadmsgcat.c:1177:9
#10 0x00007f3fd8e39e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
#11 0x00007f3fd20ed734 isKnownNonZero(llvm::Value const*, llvm::APInt const&, llvm::SimplifyQuery const&, unsigned int) ValueTracking.cpp:0:0
#12 0x00007f3fd20fb8b3 llvm::isKnownNonZero(llvm::Value const*, llvm::SimplifyQuery const&, unsigned int) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMAnalysis.so.19.0git+0x4fb8b3)
#13 0x00007f3fd20eb4c6 isKnownNonZeroFromOperator(llvm::Operator const*, llvm::APInt const&, unsigned int, llvm::SimplifyQuery const&) ValueTracking.cpp:0:0
#14 0x00007f3fd20ec6ec isKnownNonZero(llvm::Value const*, llvm::APInt const&, llvm::SimplifyQuery const&, unsigned int) ValueTracking.cpp:0:0
#15 0x00007f3fd20ebd52 isKnownNonZeroFromOperator(llvm::Operator const*, llvm::APInt const&, unsigned int, llvm::SimplifyQuery const&) ValueTracking.cpp:0:0
#16 0x00007f3fd20ec6ec isKnownNonZero(llvm::Value const*, llvm::APInt const&, llvm::SimplifyQuery const&, unsigned int) ValueTracking.cpp:0:0
#17 0x00007f3fd20fb8b3 llvm::isKnownNonZero(llvm::Value const*, llvm::SimplifyQuery const&, unsigned int) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMAnalysis.so.19.0git+0x4fb8b3)
#18 0x00007f3fd1d31346 llvm::BasicAAResult::aliasGEP(llvm::GEPOperator const*, llvm::LocationSize, llvm::Value const*, llvm::LocationSize, llvm::Value const*, llvm::Value const*, llvm::AAQueryInfo&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMAnalysis.so.19.0git+0x131346)
#19 0x00007f3fd1d31c17 llvm::BasicAAResult::aliasCheckRecursive(llvm::Value const*, llvm::LocationSize, llvm::Value const*, llvm::LocationSize, llvm::AAQueryInfo&, llvm::Value const*, llvm::Value const*) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMAnalysis.so.19.0git+0x131c17)
#20 0x00007f3fd1d34c64 llvm::BasicAAResult::aliasCheck(llvm::Value const*, llvm::LocationSize, llvm::Value const*, llvm::LocationSize, llvm::AAQueryInfo&, llvm::Instruction const*) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMAnalysis.so.19.0git+0x134c64)
#21 0x00007f3fd1d351d9 llvm::BasicAAResult::alias(llvm::MemoryLocation const&, llvm::MemoryLocation const&, llvm::AAQueryInfo&, llvm::Instruction const*) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMAnalysis.so.19.0git+0x1351d9)
#22 0x00007f3fd1d05a91 llvm::AAResults::alias(llvm::MemoryLocation const&, llvm::MemoryLocation const&, llvm::AAQueryInfo&, llvm::Instruction const*) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMAnalysis.so.19.0git+0x105a91)
#23 0x00007f3fd1d0b536 llvm::AAResults::getModRefInfo(llvm::Instruction const*, std::optional<llvm::MemoryLocation> const&, llvm::AAQueryInfo&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMAnalysis.so.19.0git+0x10b536)
#24 0x00007f3fd1f8a0df bool instructionClobbersQuery<llvm::BatchAAResults>(llvm::MemoryDef const*, llvm::MemoryLocation const&, llvm::Instruction const*, llvm::BatchAAResults&) MemorySSA.cpp:0:0
#25 0x00007f3fd1f96b81 (anonymous namespace)::ClobberWalker::findClobber(llvm::BatchAAResults&, llvm::MemoryAccess*, (anonymous namespace)::UpwardsMemoryQuery&, unsigned int&) MemorySSA.cpp:0:0
#26 0x00007f3fd1f9859b llvm::MemorySSA::ClobberWalkerBase::getClobberingMemoryAccessBase(llvm::MemoryAccess*, llvm::BatchAAResults&, unsigned int&, bool, bool) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMAnalysis.so.19.0git+0x39859b)
#27 0x00007f3fd1f99d8a llvm::MemorySSA::SkipSelfWalker::getClobberingMemoryAccess(llvm::MemoryAccess*, llvm::BatchAAResults&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMAnalysis.so.19.0git+0x399d8a)
#28 0x00007f3fd2b9d9d2 getClobberingMemoryAccess(llvm::MemorySSA&, llvm::BatchAAResults&, llvm::SinkAndHoistLICMFlags&, llvm::MemoryUseOrDef*) LICM.cpp:0:0
#29 0x00007f3fd2baea7e pointerInvalidatedByLoop(llvm::MemorySSA*, llvm::MemoryUse*, llvm::Loop*, llvm::Instruction&, llvm::SinkAndHoistLICMFlags&, bool) LICM.cpp:0:0
#30 0x00007f3fd2baf308 llvm::canSinkOrHoistInst(llvm::Instruction&, llvm::AAResults*, llvm::DominatorTree*, llvm::Loop*, llvm::MemorySSAUpdater&, bool, llvm::SinkAndHoistLICMFlags&, llvm::OptimizationRemarkEmitter*) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMScalarOpts.so.19.0git+0x1af308)
#31 0x00007f3fd2bb1992 llvm::hoistRegion(llvm::DomTreeNodeBase<llvm::BasicBlock>*, llvm::AAResults*, llvm::LoopInfo*, llvm::DominatorTree*, llvm::AssumptionCache*, llvm::TargetLibraryInfo*, llvm::Loop*, llvm::MemorySSAUpdater&, llvm::ScalarEvolution*, llvm::ICFLoopSafetyInfo*, llvm::SinkAndHoistLICMFlags&, llvm::OptimizationRemarkEmitter*, bool, bool) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMScalarOpts.so.19.0git+0x1b1992)
#32 0x00007f3fd2bb3c0d (anonymous namespace)::LoopInvariantCodeMotion::runOnLoop(llvm::Loop*, llvm::AAResults*, llvm::LoopInfo*, llvm::DominatorTree*, llvm::AssumptionCache*, llvm::TargetLibraryInfo*, llvm::TargetTransformInfo*, llvm::ScalarEvolution*, llvm::MemorySSA*, llvm::OptimizationRemarkEmitter*, bool) LICM.cpp:0:0
#33 0x00007f3fd2bb4a83 llvm::LICMPass::run(llvm::Loop&, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>&, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMScalarOpts.so.19.0git+0x1b4a83)
#34 0x00007f3fd46c66a6 llvm::detail::PassModel<llvm::Loop, llvm::LICMPass, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&>::run(llvm::Loop&, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>&, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMPasses.so.19.0git+0xc66a6)
#35 0x00007f3fd2c1c89f std::optional<llvm::PreservedAnalyses> llvm::PassManager<llvm::Loop, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&>::runSinglePass<llvm::Loop, std::unique_ptr<llvm::detail::PassConcept<llvm::Loop, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&>, std::default_delete<llvm::detail::PassConcept<llvm::Loop, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&>>>>(llvm::Loop&, std::unique_ptr<llvm::detail::PassConcept<llvm::Loop, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&>, std::default_delete<llvm::detail::PassConcept<llvm::Loop, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&>>>&, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>&, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&, llvm::PassInstrumentation&) (.isra.0) LoopPassManager.cpp:0:0
#36 0x00007f3fd2c1caf2 llvm::PassManager<llvm::Loop, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&>::runWithoutLoopNestPasses(llvm::Loop&, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>&, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMScalarOpts.so.19.0git+0x21caf2)
#37 0x00007f3fd2c1c29e llvm::PassManager<llvm::Loop, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&>::run(llvm::Loop&, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>&, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMScalarOpts.so.19.0git+0x21c29e)
#38 0x00007f3fd46c6586 llvm::detail::PassModel<llvm::Loop, llvm::PassManager<llvm::Loop, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&>, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&>::run(llvm::Loop&, llvm::AnalysisManager<llvm::Loop, llvm::LoopStandardAnalysisResults&>&, llvm::LoopStandardAnalysisResults&, llvm::LPMUpdater&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMPasses.so.19.0git+0xc6586)
#39 0x00007f3fd2c1f2ee llvm::FunctionToLoopPassAdaptor::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMScalarOpts.so.19.0git+0x21f2ee)
#40 0x00007f3fd80d4cb6 llvm::detail::PassModel<llvm::Function, llvm::FunctionToLoopPassAdaptor, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMX86CodeGen.so.19.0git+0xd4cb6)
#41 0x00007f3fd1748240 llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.19.0git+0x348240)
#42 0x00007f3fd80d3a56 llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMX86CodeGen.so.19.0git+0xd3a56)
#43 0x00007f3fd1746ed7 llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.19.0git+0x346ed7)
#44 0x00007f3fd80d4e56 llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMX86CodeGen.so.19.0git+0xd4e56)
#45 0x00007f3fd17450f1 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.19.0git+0x3450f1)
#46 0x00007f3fd98f9314 llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::PassPlugin>, llvm::ArrayRef<std::function<void (llvm::PassBuilder&)>>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool, bool) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMOptDriver.so.19.0git+0x2d314)
#47 0x00007f3fd9906502 optMain (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMOptDriver.so.19.0git+0x3a502)
#48 0x00007f3fd8e29d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#49 0x00007f3fd8e29e40 call_init ./csu/../csu/libc-start.c:128:20
#50 0x00007f3fd8e29e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#51 0x00005e3711de5095 _start (bin/opt+0x1095)
Aborted (core dumped)
```

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 2]
```diff
@@ -2632,6 +2632,13 @@ static bool isKnownNonZeroFromOperator(const Operator *I,
             Q.DL.getTypeSizeInBits(I->getType()).getFixedValue())
       return isKnownNonZero(I->getOperand(0), Q, Depth);
     break;
+  case Instruction::Trunc: {
+    auto *TI = cast<TruncInst>(I);
+    // nuw/nsw trunc preserves zero/non-zero status of input.
+    if (TI->hasNoSignedWrap() || TI->hasNoUnsignedWrap())
+      return isKnownNonZero(TI->getOperand(0), Q, Depth + 1);
```
[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp:2639]: https://github.com/llvm/llvm-project/blob/d5093aac543dfd24f339f7d43aa02282f7c16fa5/llvm/lib/Analysis/ValueTracking.cpp#L3075-L3077

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 3]
```diff
@@ -2632,6 +2632,13 @@ static bool isKnownNonZeroFromOperator(const Operator *I,
             Q.DL.getTypeSizeInBits(I->getType()).getFixedValue())
       return isKnownNonZero(I->getOperand(0), Q, Depth);
     break;
+  case Instruction::Trunc: {
+    auto *TI = cast<TruncInst>(I);
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:2636]: This probably needs to be dyn_cast, as trunc can occur in constant expressions.

**Comment by dtcxzyw** [comment]
I will re-test this patch later :)

**Review by dtcxzyw — APPROVED** [review]
LGTM.
